File: app/src/main/java/net/j4dy/familypicingame/games/common/Vector2D.kt
package net.j4dy.familypicingame.games.common

data class Vector2D(val x: Float = 0f, val y: Float = 0f) {
    operator fun plus(other: Vector2D) = Vector2D(x + other.x, y + other.y)
    operator fun minus(other: Vector2D) = Vector2D(x - other.x, y - other.y)
    operator fun times(scalar: Float) = Vector2D(x * scalar, y * scalar)
    operator fun div(scalar: Float) = if (scalar != 0f) Vector2D(x / scalar, y / scalar) else Vector2D(0f, 0f)
    
    fun length() = Math.sqrt((x * x + y * y).toDouble()).toFloat()
    
    fun normalize(): Vector2D {
        val len = length()
        return if (len > 0f) this / len else Vector2D(0f, 0f)
    }
    
    fun dot(other: Vector2D) = x * other.x + y * other.y
    
    fun distance(other: Vector2D) = (this - other).length()
    
    fun limit(max: Float): Vector2D {
        val len = length()
        return if (len > max) this.normalize() * max else this
    }
}

File: lib/games/common/vector_2d.dart
import 'dart:math' as math;

class Vector2D {
  final double x;
  final double y;

  const Vector2D(this.x, this.y);

  Vector2D operator +(Vector2D other) => Vector2D(x + other.x, y + other.y);
  Vector2D operator -(Vector2D other) => Vector2D(x - other.x, y - other.y);
  Vector2D operator *(double scalar) => Vector2D(x * scalar, y * scalar);
  Vector2D operator /(double scalar) => scalar != 0.0 ? Vector2D(x / scalar, y / scalar) : const Vector2D(0.0, 0.0);

  double length() => math.sqrt(x * x + y * y);

  Vector2D normalize() {
    final len = length();
    return len > 0.0 ? this / len : const Vector2D(0.0, 0.0);
  }

  double dot(Vector2D other) => x * other.x + y * other.y;

  double distance(Vector2D other) => (this - other).length();

  Vector2D limit(double max) {
    final len = length();
    return len > max ? normalize() * max : this;
  }
}
